### PR TITLE
Adding support for Heidi Climate System

### PR DESCRIPTION
**Pull request by jmwaldrip**
This is for the Climate360 bed with active heating and cooling in the entire bed.

### CODE REVIEW

**Comment by kbickar** [comment]
What does this look like in the app? Might just call it "Climate" instead of Heidi.

Can you also create a base class that this can inherit from (Just in case that's available for another model)

**Comment by jmwaldrip** [comment]
Yeah, I went back and forth a couple times. I landed on Heidi because it was the name of the api code. 

Here is a screenshot of the app. 

image

It's called "Core". 

Yeah, I can add a base.

**Comment by jmwaldrip** [comment]
For the base class, not sure how much you're going to want in it. But it won't implement anything other than init probably since I don't have any other API patterns other than fuzion.

File: asyncsleepiq/fuzion/heidi.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        """Update the heidi data through the API."""
+        args = [SIDES_FULL[self.side].lower()]
+        self.preset = await self._api.bamkey(self.bed_id, "GetHeidiMode", args)
+
+    def __str__(self) -> str:
+        """Return string representation."""
+        return f"SleepIQHeidi[{self.side}]: {'On' if self.is_on else 'Off'}, {self.timer}, {HeidiTemps(self.temperature).name}"
+    __repr__ = __str__
+
+    async def turn_on(self, temperature: HeidiTemps, time: int)  -> None:
```
Comment (kbickar) [inline asyncsleepiq/fuzion/heidi.py:47]: This can go in base class

File: asyncsleepiq/fuzion/heidi.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        """Turn on heidi mode through API."""
+        await self.set_heidi_mode(temperature, time)
+
+    async def turn_off(self) -> None:
+        """Turn off heidi mode through API."""
+        # The API requires a valid time value even if we're turning the warmer off
+        await self.set_heidi_mode(HeidiTemps.OFF, 1)
+
+    def time_key(self):
```
Comment (kbickar) [inline asyncsleepiq/fuzion/heidi.py:56]: Not used

File: asyncsleepiq/fuzion/heidi.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        return 'heidiTimer' + SIDES_FULL[self.side]
+
+    def temp_key(self, read = False):
```
Comment (kbickar) [inline asyncsleepiq/fuzion/heidi.py:59]: Not used

File: asyncsleepiq/fuzion/heidi.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        args = [SIDES_FULL[self.side].lower(), temperature.name.lower(), str(time)]
+        await self._api.bamkey(self.bed_id, "SetHeidiMode", args)
+        await self.update({})
+
+    async def update(self, data: dict[str, Any]) -> None:
+        """Update the heidi data through the API."""
+        args = [SIDES_FULL[self.side].lower()]
+        self.preset = await self._api.bamkey(self.bed_id, "GetHeidiMode", args)
+
+    def __str__(self) -> str:
```
Comment (kbickar) [inline asyncsleepiq/fuzion/heidi.py:42]: Can go in base class

File: asyncsleepiq/fuzion/foundation.py [hunk 5]
```diff
@@ -111,6 +115,14 @@ async def init_foot_warmers(self) -> None:
             if result == "1":
                 self.foot_warmers.append(SleepIQFuzionFootWarmer(self._api, self.bed_id, side, 0, 0))
 
+    async def init_heidi_modes(self) -> None:
+        """Initialize list of heidi modes available on foundation."""
+        for side in [Side.LEFT, Side.RIGHT]:
+            print(side)
```
Comment (kbickar) [inline asyncsleepiq/fuzion/foundation.py:121]: Remove print statement

File: asyncsleepiq/fuzion/foundation.py [hunk 6]
```diff
@@ -111,6 +115,14 @@ async def init_foot_warmers(self) -> None:
             if result == "1":
                 self.foot_warmers.append(SleepIQFuzionFootWarmer(self._api, self.bed_id, side, 0, 0))
 
+    async def init_heidi_modes(self) -> None:
+        """Initialize list of heidi modes available on foundation."""
+        for side in [Side.LEFT, Side.RIGHT]:
+            print(side)
+            result = await self._api.bamkey(self.bed_id, "GetHeidiPresence", args=[SIDES_FULL[side].lower()])
+            if result == "true":
```
Comment (kbickar) [inline asyncsleepiq/fuzion/foundation.py:123]: Does this return "true" or "1"?

File: asyncsleepiq/consts.py [hunk 7]
```diff
@@ -21,6 +21,14 @@ class FootWarmingTemps(int, enum.Enum):
     MEDIUM = 57
     HIGH = 72
 
+class HeidiTemps(int, enum.Enum):
```
Comment (kbickar) [inline asyncsleepiq/consts.py:24]: CoreTemps

File: asyncsleepiq/fuzion/heidi.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        """Initialize heidi object."""
+        self._api = api
+        self.bed_id = bed_id
+        self.side = side
+        self.is_on = (temperature > 0)
+        self.timer = timer
+        self.temperature = temperature
+
+
+    async def set_heidi_mode(self, temperature: HeidiTemps, time: int) -> None:
```
Comment (kbickar) [inline asyncsleepiq/fuzion/heidi.py:28]: maybe just `set_mode` as the name

**Review by kbickar — CHANGES_REQUESTED** [review]
Added some comments.

Looking at the app screenshot, it might make sense to be called "CoreClimate" or something like that.  The library shouldn't be exporting classes/functions with "heidi" in the name since from the outside no one knows what that means.

File: asyncsleepiq/fuzion/foundation.py (hunk 6 above)
Comment (jmwaldrip) [reply asyncsleepiq/fuzion/foundation.py:123]: It returns "true" not "1". 

I found that during testing.

File: asyncsleepiq/fuzion/heidi.py (hunk 3 above)
Comment (jmwaldrip) [reply asyncsleepiq/fuzion/heidi.py]: Removed

File: asyncsleepiq/fuzion/heidi.py (hunk 4 above)
Comment (jmwaldrip) [reply asyncsleepiq/fuzion/heidi.py]: Moved to base class

File: asyncsleepiq/fuzion/foundation.py (hunk 5 above)
Comment (jmwaldrip) [reply asyncsleepiq/fuzion/foundation.py]: removed

File: asyncsleepiq/fuzion/heidi.py (hunk 8 above)
Comment (jmwaldrip) [reply asyncsleepiq/fuzion/heidi.py]: changed to set_mode

File: asyncsleepiq/consts.py (hunk 7 above)
Comment (jmwaldrip) [reply asyncsleepiq/consts.py]: Changed to CoreTemps

**Comment by jmwaldrip** [comment]
Thanks for the feedback! I updated the pull request with the requested changes.

File: asyncsleepiq/core_climate.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+from __future__ import annotations
+
+from typing import Any
+
+from .api import SleepIQAPI
+
+from .consts import SIDES_FULL, CoreTemps, Side
+
+
+class SleepIQCoreClimate:
```
Comment (kbickar) [inline asyncsleepiq/core_climate.py:11]: Export this in the `__init__.py`

File: asyncsleepiq/fuzion/core_climate.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+class SleepIQFuzionCoreClimate(SleepIQCoreClimate):
+    """
+        CoreClimate (Also known as Heidi) representation for SleepIQ Fuzion API.
+        Heidi is the name of the climate calls in the SleepIQ API.
+        Controls heating and cooling.
+    """
+
+    async def set_mode(self, temperature: CoreTemps, time: int) -> None:
+        """Set core climate state through API."""
+        if time <= 0 or time > 360:
```
Comment (kbickar) [inline asyncsleepiq/fuzion/core_climate.py:22]: Do the same as with the foot warmer (on main branch) and have the max time be a class variable

File: asyncsleepiq/core_climate.py (hunk 9 above)
Comment (jmwaldrip) [reply asyncsleepiq/core_climate.py:11]: Exported

File: asyncsleepiq/fuzion/core_climate.py (hunk 10 above)
Comment (jmwaldrip) [reply asyncsleepiq/fuzion/core_climate.py]: Good call, updated it.

**Review by kbickar — APPROVED** [review]

**Pull request merged**